### PR TITLE
Add MVC support for reference array copy

### PR DESCRIPTION
**Pull request by fjeremic**
Currently the z Systems codegen is generating a load-store loop for
forward reference arraycopy because of a historic bug that is now moot.
This commit enables generation of the same MVC sequence for reference
arraycopy that we use for primitive arraycopy.

We take this time to reorganize, simplify, and document all arraycopy
evaluators on z Systems. We now have a consolidated path for primitive
and reference arraycopy with maximal code reusage.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

### CODE REVIEW

**Comment by fjeremic** [comment]
@dsouzai could you please review?

**Comment by mgaudet** [comment]
sp. not moot -> now moot. :D

**Comment by charliegracie** [comment]
Are the reference array copies still happening a `word` at a time?

**Comment by fjeremic** [comment]
> sp. not moot -> now moot. :D

Fixed.

> Are the reference array copies still happening a word at a time?

Quoting _Principles of Operation_ Section 5-125:

> When destructive overlap does not exist, the operands of MOVE (**MVC**), MOVE WITH KEY, MOVE TO
PRIMARY, and MOVE TO SECONDARY are accessed as follows:
>1. The first operand is accessed in a left-to-right direction, and all bytes accessed within a double 
word appear to be accessed concurrently as observed by other CPUs.
>2. The second operand is accessed in a left-to-right direction, and all bytes within a doubleword in the second operand that are moved into a single doubleword in the first operand appear to be fetched concurrently as observed by other CPUs. Thus, if the first and second operands begin on the same byte offset within a doubleword, the fetch of the second operand appears to be doubleword concurrent as observed by other CPUs. If the offsets within a doubleword differ by 4, the fetch of the second operand appears to be word concurrent as observed by other CPUs.

So yes MVC is block concurrent on a **double-word** boundary.

**Comment by charliegracie** [comment]
Thanks for the info! Just making sure concurrent GCs do not see incorrect values during concurrent marking.

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -13461,454 +13461,133 @@ OMR::Z::TreeEvaluator::arraysetEvaluator(TR::Node * node, TR::CodeGenerator * cg
...
-   if (copyByteDstReg)
-      {
-      cg->stopUsingRegister(byteDstReg);
-      }
-   if (copyByteLenReg)
+   else
       {
-      cg->stopUsingRegister(byteLenReg);
+      // TODO: There are cases under which the following assert will trigger. These need to be fixed and this assert needs to be re-enabled.
+      // TR_ASSERT(node->getArrayCopyElementType() == TR::Address, "Reference arraycopy element type should be a TR::Address but was '%s'", node->getArrayCopyElementType().toString());
```
Comment (dsouzai) [inline compiler/z/codegen/OMRTreeEvaluator.cpp]: It seems like this assert could be re-enabled now that we can only get here if it's a ref array copy but I'm not 100% sure about that. Worth making a note of this anyway.

**Comment by dsouzai** [comment]
Looks fine to me. Probably worth having Joran review this for a better functional correctness review (I don't know what his GitHub handle is...).

File: compiler/z/codegen/OMRTreeEvaluator.cpp (hunk 1 above)
Comment (fjeremic) [inline compiler/z/codegen/OMRTreeEvaluator.cpp]: This is not correct. Currently for example in VP we generate reference arraycopy with `TR::Int32` data type in some cases. This change is outside the scope of this PR hence why I left it as a TODO.

**Comment by fjeremic** [comment]
> Looks fine to me. Probably worth having Joran review this for a better functional correctness review (I don't know what his GitHub handle is...).

@joransiu

**Comment by fjeremic** [comment]
@joransiu any concerns or suggestions regarding this commit?

**Comment by joransiu** [comment]
LGTM.  looks a lot cleaner now. :+1:

**Comment by fjeremic** [comment]
@0xdaryl any suggestions from you?
